feat: Cloud storage + Updated GitHub oauth

KyleleeSea opened the pull request:
# Description
(Branch name is misleading, this is for the entire cloud storage system.)

Currently, diagrams are saved only on local storage. This makes it difficult for a user to edit diagrams cross platform or cross browser. As a solution, this PR implements cloud storage for saved diagrams. We choose Firestore as our cloud storage provider for cost, auth integration with security rules, and performance reasons. Additionally, we rework the GitHub oauth method to be more secure to support our new system. 

# Implementation strategy and design decisions
### Authentication
Previously there were two main issues with the authentication strategy:
1. User information would be stored and and queried from local storage. This creates the a major security vulnerability: A user could simply overwrite their username in local storage to "spoof" being another user. 
2. Rather than making requests to the GitHub oauth API directly, the server redirected to a vercel hosted serverless function which made the requests (this was because of a cors policy + SharedArrayBuffers in the web worker which is no longer a problem).

The new Authentication mechanism is simpler and secure. We continue using GitHub oauth, but we handle it through Firebase auth which removes the need for the additional vercel serverless function and provides a user object which can be securely queried. 

In order to utilize this Firebase auth function, we had to change the server headers policy in vite.config.ts to:
` headers: { "Cross-Origin-Opener-Policy": "same-origin-allow-popups" }`
Originally, it was `"Cross-Origin-Embedder-Policy": "require-corp", "Cross-Origin-Opener-Policy": "same-origin"`
The new setting allows the application to open popups with different domain URLs, which is necessary to use the Firebase handled oauth menu. 
Previously this was not possible because the "same-origin" policy was required for the SharedArrayBuffer to be used in the web worker, which has since been removed.
[A "same-origin" policy is usually implemented for security reasons](https://portswigger.net/web-security/cors/same-origin-policy). However, I do not see this as an issue in our application as there is nowhere a user can insert runnable Javascript into our application (the editor is sanitized). 

### Database Structure
-- Top level
----> userid indexed collections
--------> user documents (fields: diagramId, editorVersion, lastModified, name, style, substance, domain)

### Security Rules
Users only have access to read, write, update, and create to diagrams in collections (folders) with title equal to their userid. We check userid using Firebase auth functions which require the user be logged in via the GitHub oauth method previous described. Security rules enforced via Firebase security rules system which automatically integrates with Firebase auth.

### Saved Tab & Caching 
If user is not logged in, this component displays "Please sign in to use saved diagrams!" and a Sign In button.
This tab is populated by the following method: On the first render of the application and on updates to the user state (i.e. if the user logs in), if the user is logged in, then a GET request is made to Firestore to retrieve all diagrams in the user's collection. This data is then cached in a local state "savedFilesState" and used to populate the saved tab. The saved tab then references savedFilesState which is updated on future operations. This method is designed to minimize GET requests made to Firestore, as GET requests are only made once when a user enters a logged in state. 

### Saved Tab Ordering
Additionally, the saved tab is now ordered by lastModified time (so the most recently modified diagram will appear at the top).

### Diagram Loading
When a user clicks a diagram name in the saved tab, the savedFilesState is consulted for the corresponding diagramId. If found in local state, the diagram is loaded. Otherwise, a request is made to Firestore. The diagram should be found in all cases as the local state is updated appropriately with each operation, the fall back to Firestore request exists in the case of some unforeseen edge case. 

### Diagram Saving
Autosaving every 500ms is no longer viable as Firestore charges per write request made (after a free quota is used up), regardless of how small the write is. As a result, we now added the following:
1. Autosave every 5 seconds after a user has finished editing
2. Added cmd+S and ctrl+S as save shortcuts 

When a saved diagram is updated its state is updated in savedFilesState and then the changes are pushed to Firestore.

When a new diagram is saved for the first time, its state is added to savedFilesState and then a new document in the user's collection is made for the diagram. Duplicate works similarly. 

Editing a saved diagram name will push the change to the cloud upon exit of the input box (clicking away, escape key, or enter key).

### Merge Conflict Resolution
Suppose a user has diagram and edits it on two devices. They open their diagram on devices A and B. On device A, they edit the diagram and save. These saves are not reflected in what's shown on device B. Now on device B they make an edit and attempt to save. Currently the behavior is that the system recognizes there are changes made by device A (by checking lastmodified date on cloud with local copy) and notifies user that there is a merge conflict and asks if they want to override. 

A stretch goal would be to add a built-in merge conflict resolver like what GitHub has. 

### Delete
Works as one would expect. Removes file from local state and then removed from cloud database. Slightly slower than local storage, now delete takes about 1 second to complete. 

### Download Diagram
Does not interact with the cloud system and consults local state. This is because for a user to click download they must have that diagram loaded, in which case the diagram is in the currentWorkspace local state. 

### Diagram Sharing
Share button now only appears if diagram is saved. 
Previously if a diagram was unsaved and share was clicked, the diagram would automatically save. The newest save function cannot be called directly from the usePublishGist function because of React callback restrictions (the old save function wasn't a React callback). Thus our options were as follows:
1. Duplicate save code into usePublishGist 
2. Share button only appearing if diagram is saved
3. If diagram is unsaved and share clicked warn user, asking if they'd like to proceed 
I thought option 2 was the best as it also makes the UI cleaner with less clutter in the top bar. I am open to suggestions though. 

# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have reviewed any generated registry diagram changes

# Open questions
1. The autosave after every 5 seconds has two minor bugs, neither of which cause any substantive problems. First, the first edit made after a diagram is loaded in does not trigger the autosave effect. This is because of nuances with React state updates. Second, When a new diagram is loaded in the autosave effect is called. If no edits are made, nothing will happen though. This is because of limitations in checking previous vs current workspace state. 

2. Also autosave is currently annoying if there exists a merge conflict. Since autosave calls the save function, if there is a merge conflict, the user will get a notification every 5 seconds after editing asking if they want to overwrite changes with a merge. Our options are: A) Ignore this for now and leave as is. B) Disable autosave when there's a merge conflict (this will require a decent deal of added complexity to the codebase). 

3. This is tangentially related as it pertains to both the old and new system. GitHub access token is still being stored in local storage (old system stored it in local storage as well). The access token is necessary to publish Gists with the share button. The access token is not stored in the Firebase user object and only returned on initial GitHub oauth login. We store it in local storage to persist the token across sessions so a user doesn't need to oauth login on every new session. However, hypothetically a user could change their local storage access token to another user's access token to publish Gists as another user. I don't see this as a concern as they'd have to guess another user's access token which would be a problem with oauth in general, not really our concern. Again, this token was saved in local storage in the old system as well.

4. Error messages for login, logout, save, etc are currently not very descriptive. I've only added generic "Error [x action]" since the default Firebase errors specify we're using Firebase. 

### Notes
The Files Changed bar is a bit misleading, I manually merged in Griffin's changes so I could add github oauth as previously SharedArrayBuffer conflicted. So there's some diffs shown about Error.ts and compilation that are not my changes and are already part of the main branch. 

We're still waiting on the Penrose Firebase account, currently this is deployed on my personal account.

Github oauth callback url is set to localhost:3000, so please test on localhost, github oauth will not work on cloudflare pages. (Will change before merge)

wodeni commented on package-lock.json:
As we are using Yarn, there shouldn't be a `package-lock.json` file. Do you mind deleting it?

wodeni commented on packages/editor/vite.config.ts:
```diff
@@ -18,10 +18,7 @@ export default defineConfig({
+    headers: { "Cross-Origin-Opener-Policy": "same-origin-allow-popups" },
   },
   preview: {
     headers: {
```
the same policies probably apply for the previews (i.e. local production builds for testing) right?

KyleleeSea replied:
would previews be the cloudflare pages that get deployed prior to what's shown in production merge in our case?

wodeni commented on packages/editor/src/state/callbacks.ts:
```diff
@@ -363,20 +405,12 @@ export const useDownloadPdf = () =>
-    !workspace.metadata.location.saved
+    (workspace.metadata.location.kind === "stored" &&
+      !workspace.metadata.location.saved) ||
+    workspace.metadata.location.kind == "local"
```
Just making sure: this means that any unsaved non-example workspace is considered "clean"?

KyleleeSea replied:
Example workspaces (both edited and unedited) are considered clean
Saved workspaces with no unsaved changes are considered clean

Unsaved workspaces (i.e. if you click "new workspace" work in it and don't save it ever) are considered unclean (this is the workspace.meta.location.kind == "local" check)
Unsaved workspaces marked as "stored" that have unsaved changes are considered unclean

Does this clarify?

wodeni commented on packages/editor/src/utils/firebaseUtils.tsx:
```diff
@@ -0,0 +1,163 @@
+import { useRecoilCallback } from "recoil";
+import { SavedWorkspaces, Workspace, settingsState } from "../state/atoms.js";
+
+const firebaseConfig = {
```
we discussed offine: it might be good practice to put this in the CI/CD environment instead.

wodeni commented on packages/editor/src/utils/firebaseUtils.tsx:
```diff
@@ -0,0 +1,163 @@
+        }
+      })
+      .catch((error) => {
+        toast.error("Error logging in");
```
```suggestion
        toast.error(`Error logging in: ${error}`);
```
if the error messages are meaningful, it'd still be good to propagate them to the user for reporting.

wodeni commented on packages/editor/src/utils/firebaseUtils.tsx:
```diff
@@ -0,0 +1,163 @@
+        toast.success("Logged out");
+      })
+      .catch((error) => {
+        toast.error("Error logging out");
```
same here: propagate error message.

wodeni commented on packages/editor/src/utils/firebaseUtils.tsx:
```diff
@@ -0,0 +1,163 @@
+): Workspace {
+  return {
+    metadata: {
+      name: name,
+      lastModified: lastModified,
+      id: id,
+      editorVersion: editorVersion,
```
```suggestion
      name,
      lastModified,
      id,
      editorVersion,
```

shorthand to reduce noise

wodeni commented on packages/editor/src/utils/firebaseUtils.tsx:
```diff
@@ -0,0 +1,163 @@
+        docData.domain,
+      );
+    } else {
+      toast.error("Could not fetch diagram");
```
```suggestion
      toast.error(`Could not fetch diagram ${name} from cloud storage because it doesn't exist.`);
```

KyleleeSea replied:
Updated to error message: "Could not fetch diagram from cloud storage because it doesn't exist."
getDiagram would be called from loadWorkspace, where name isn't defined. This is called using diagramId as an identifier which isn't meaningful to the user (It's a randomly generated string)

wodeni commented on packages/editor/src/main.tsx:
```diff
@@ -20,14 +20,16 @@ class ErrorBoundary extends Component<{ children: ReactElement }> {
-    </RecoilRoot>
-  </React.StrictMode>,
+  // React Strict mode is useful for development but causes an additional
+  // render which causes bugs with CheckURL firing twice from App useEffect
```
Based on the [docs](https://react.dev/reference/react/StrictMode), strict mode doesn't affect prod builds. Okay to comment this out IMO.

wodeni commented on packages/editor/src/state/callbacks.ts:
```diff
@@ -707,20 +714,166 @@ export const useDeleteLocalFile = () =>
+                currentWorkspaceState,
+              ).contents;
+              set(savedFilesState, (savedFiles) => {
+                const { [id]: removedFile, ...newFiles } = savedFiles;
```
huh didn't know about destructuring assignment and rest properties ([docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#rest_property)). pretty cool!

wodeni commented on packages/editor/src/state/callbacks.ts:
```diff
@@ -707,20 +714,166 @@ export const useDeleteLocalFile = () =>
+        style: currentWorkspace.files.style.contents,
+        domain: currentWorkspace.files.domain.contents,
+      })
+        .catch((error) => toast.error("Encountered an error"))
```
log the error here if it's meaningful?

KyleleeSea replied:
Error message and error code now reported in console log. 

Error messages not displayed to console because some are preceded by "Firebase: [message]" which felt ugly in the UI.

wodeni commented on packages/editor/src/components/SavedBrowser.tsx:
```diff
@@ -1,45 +1,176 @@
+          currentWorkspaceState,
+        ).contents;
+        if (repeat) return;
+        // Cmd+s or Cntrl+s
```
```suggestion
        // Cmd+s or Ctrl+s
```

wodeni commented on packages/editor/src/components/SavedBrowser.tsx:
```diff
@@ -1,45 +1,176 @@
+ * Due to current layout scheme, this runs no matter what side tab the user
+ * is on. If layout scheme ever changes, this may need to be moved to App
+ */
+const autosaveHook = () => {
```
I repro'ed another case when autosave doesn't run:
* Load the IDE with wide layout (default left tab === "examples")
* Do not touch "saved" and just start editing
* No autosave (maybe expected)
* Manual save by Cmd+S
* Wait... and no autosave triggered

My theory is that `flexlayout` actually doesn't mount the component until it's expanded. Therefore, if the user never touches the tab, the hooks would never fire. 

My suggestion is to move the hook calls to `App` instead.

wodeni commented on packages/editor/src/components/SavedBrowser.tsx:
```diff
@@ -1,45 +1,176 @@
+              )}
+          </div>
+        </div>
+      ) : (
```
The UI is a bit bold :P. We talked off line about doing an actual overhaul.

KyleleeSea commented:
Updates from this most recent commit:
1) Autosave triggers consistently (Bug due to a race with markUnsaved atom effect)
2) cmd+s/ctrl+s will save "local" (what u get when u click new workspace) workspaces to the saved browser. 
3) If user editing a saved workspace, cmd+s and autosave now works even if user has not opened the saved browser. (To replicate: open editor without navigating to saved browser, cmd+s, edit, observe autosave and cmd+s do things now)
4) Meaningful error messages logged to console

======
Some notes...
Autosave couldn't be moved to App. For some reason the call to get value of currentWorkspaceState in App made the editor VERY laggy. I think this has to do with it triggering rerenders since the editor pane is part of App and uses currentWorkspaceState. The issue found about autosave not running was indeed with it being in savedBrowser and savedBrowser not loading unless clicked to. As such, we moved autosave (and saveshortcut since it's associated) into TopBar. This feels a bit hacky and suboptimal, but after trying different solutions for a few hours, this seemed like the best solution.

Need to think more about the tradeoff of moving firebase config into cloudflare env

KyleleeSea commented:
Latest commit fixed two more issues:
1) On edit of an example workspace, it will now be marked as unclean and the save button in the top bar will show + save shortcut will save it. 
Old behavior:
https://github.com/penrose/penrose/assets/85892844/985ec561-35d6-4ad5-9a91-c8c65a1e42b1

New behavior:
https://github.com/penrose/penrose/assets/85892844/66dfa34e-cb36-4395-bdaf-26903aacb0f0


2) Fixed issue where layout select didn't work
Old behavior:
https://github.com/penrose/penrose/assets/85892844/2bc9bbc8-5630-4e2b-99f7-44def8484515

Fixed behavior:
https://github.com/penrose/penrose/assets/85892844/8a001714-83e4-44bb-89b7-cd615d4831bb

KyleleeSea commented:
Fixed issue where error notification would appear in race between autosave and delete.

Old behavior:
https://github.com/penrose/penrose/assets/85892844/95123607-7b8d-4025-92e4-4b8eb1b90971

New behavior:
https://github.com/penrose/penrose/assets/85892844/0e2c3cf1-c609-469a-818b-dcc38a52282b

wodeni reviewed (changes requested):
looks like `package-lock.json` is there. other than that everything looks good!

wodeni reviewed (approved):
LGTM!
